tests/tests/many_threads.c

```diff
@@ -1,51 +1,83 @@
-// ignore: broken during new control point design
-// Compiler:
 // Run-time:
+//   env-var: YKD_PRINT_IR=jit-pre-opt,aot
+//   env-var: YKD_SERIALISE_COMPILATION=1
+//   env-var: YKD_PRINT_JITSTATE=1
+//   stderr:
+//     ...
+//     jit-state: enter-jit-code
+//     ...
 
 // Check that compiling and running traces in parallel works.
+//
+// Due to non-determinism, it's difficult to devise a meaningful expected
+// outcome. Although threads will block waiting for the trace compiler (due to
+// YKD_SERIALISE_COMPILATION=1), there is still non-determinism in which
+// threads execute compiled trace code and how they might interleave.
+//
+// For this reason, this test only really checks that nothing crashes.
 
 #include <assert.h>
 #include <err.h>
 #include <pthread.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <yk.h>
 #include <yk_testing.h>
 
 #ifdef linux
 #include <sys/sysinfo.h>
 #endif
 
-static void *trace(void *unused) {
-  for (int i = 0; i < 3; i++) {
-    __yktrace_start_tracing(HW_TRACING);
-    int res = 1 + 1;
-    void *tr = __yktrace_stop_tracing();
-    assert(res == 2);
-
-    void *ptr = __yktrace_irtrace_compile(tr);
-    __yktrace_drop_irtrace(tr);
-    void (*func)() = (void (*)())ptr;
-    func();
+#define ITERS 1000
```

Comment:
We probably need to set this much higher, otherwise it's quite possible that a thread will complete before any others have started. I'd probably set it to 100,000.

Reply:
FWIW, in my testing I didn't see any interleaving with 1000 iters, but did at 100,000. I didn't up it because a) it makes the test slow, and b) we can't test interleaving anyway.

Reply:
seconds, yes.

Reply:
Seconds is fine -- and I bet you if you remove the `printf` you're down to tiny fractions of a second.